### ISSUE TITLE
Add ability to add a point on either side of a linear element

### ISSUE THREAD

**Issue by xixixao** [excalidraw/excalidraw#1895]
I thought this was already in place, but I can't make it work:

I expect that if I select a "starting" point, then press "alt", I should be able to "prepend" points to the line. But it seems that I can only add points to the end.

<img width="401" alt="Screen Shot 2020-07-08 at 21 19 07" src="https://user-images.githubusercontent.com/1473433/86996547-d0a95180-c160-11ea-9342-365defb3c57f.png">

**Comment by dwelle** [comment]
Could be easily implemented, but what about when you select points in the middle? Should it append to one of the ends, or should it create a new point in between the selected + next point?

**Comment by xixixao** [comment]
I think creating a new point in between would be nice, the problem is, which side of the selected point? And this will leave one segment that you cannot add points to (say the first segment from p0 to p1).

This is a solved problem in other software, here's what Sketch does:

1) Doesn't require `alt` pressed, always is in "add point at front mode"
2) Allows you to highlight any segment, if you click on it you create a point at the precise location you're pointing at
3) To add point before first point, first select the first point, then you get to create a point there

I kinda like the `alt` requirement though, because without it the vector tools always feel like you create accidental points which you didn't intend to create.

**Comment by xixixao** [comment]
Here's an option (but might not be super obvious):

1) Support selecting multiple points using shift (this should be a separate issue that we don't atm)
2) If two points are selected, pressing alt creates a new one between them

One thing about this `alt` based creation: You get to see the result before you confirm it with a tap.

(there should also be a way to close the loop, not sure whether we have an issue for that)

**Comment by dwelle** [comment]
> if two points are selected, pressing alt creates a new one between them

I think this is too complex, both conceptually (hard to explain this behavior in a hint) and ergonomically.

How about this:

Create the next point based on cursor position relative to the selected point based on what "side" the cursor is on. To figure out what side to create it on, we'd consider the selected point's normal (line perpendicular to its tangent):

![image](https://user-images.githubusercontent.com/5153846/87153839-63a8cf80-c2b8-11ea-8570-68ae29c33ffa.png)

This should work for first and last point, too. Not sure if there are edge cases for which it would not work.

> (there should also be a way to close the loop, not sure whether we have an issue for that)

Actually, if you create/position the last point close to the first point, it auto-closes the loop already. The only problem is it currently doesn't show the preview of this behavior while you hold ALT.

**Comment by xixixao** [comment]
> Create the next point based on cursor position relative to the selected point based on what "side" the cursor is on by taking into account the selected point's normal (line perpendicular to its tangent):

> That said, for edge points this algo might not be desired. E.g. consider the below case:

I think this problem occurs for the middle points too (ie, I might want to create the middle point on the other side of the "normal"), and this is why the suggested approach might work (or might not be "ideal"). But it's worth implementing and seeing how it feels, maybe iterating on it later.

> Actually, if you create/position the last point close to the first point, it auto-closes the loop already. The only problem is it currently doesn't show the preview of this behavior while you hold ALT.

We should file an issue for this (if not already).

**Comment by vjeux** [comment]
@dwelle fyi, what you are talking about is called a Voronoi diagram. Each shaded area corresponds to all the pixels that are closest to a particular point in the 2d space.

**Comment by dwelle** [comment]
> I think this problem occurs for the middle points too (ie, I might want to create the middle point on the other side of the "normal"), and this is why the suggested approach might not be "ideal". But it's worth implementing and seeing how it feels, maybe iterating on it later.

That's true, but an argument can be made that more often than not when creating a point between to other points, you just want to correct for something and not deviate too far. Thus the algo I propose may be "good enough" for the common case. But it'd need to be implemented so we can try it out, as you say.

> @dwelle fyi, what you are talking about is called a Voronoi diagram. Each shaded area corresponds to all the pixels that are closest to a particular point in the 2d space.

Right, I now recall this. It's a similar concept but my proposal is not aiming to figure out which of the two points are closer to the cursor, instead I want to figure out which of the two planes (delineated by the selected point's normal) the cursor is on. In Voronoi diagrams you can notice that the borders do not intersect the points, while in my solution they do.
